### DISCUSSION TITLE
Restore NuGet package created in another repo

### DISCUSSION THREAD

**Discussion by ghost** [community/community#26856] [Packages]
<p>I’m trying to push NuGet packages to a GitHub package feed but currently can’t restore GitHub NuGet packages during the build that were created from Actions in another repo:</p>
<blockquote>
<p>Your request could not be authenticated by the GitHub Packages service. Please ensure your access token is valid and has the appropriate scopes configured.</p>
</blockquote>
<p>NuGet auth token is set with <code>actions/setup-dotnet</code>:</p>
<pre><code class="lang-yaml">      # https://github.com/actions/setup-dotnet
      -
        name: Setup .NET
        uses: actions/setup-dotnet@v1
        with:
          dotnet-version: '3.1.x'
          source-url: ${{ env.NUGET_URL }}
        env:
          NUGET_AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
</code></pre>
<p>I also tried using the same token used when pushing the packages that are attempting to be restored here but it’s the same result.</p>
<p>I’m assuming I need to modify NuGet.config so it has the credentials like:</p>
<pre><code class="lang-xml">  &lt;packageSourceCredentials&gt;
        &lt;github&gt;
            &lt;add key="Username" value="${USER}" /&gt;
            &lt;add key="ClearTextPassword" value="${PAT}" /&gt;
        &lt;/github&gt;
    &lt;/packageSourceCredentials&gt;
</code></pre>
<p>However I thought that wouldn’t be required within the context of GitHub actions in the same org?</p>
<p>If it is required, what’s the practical way of doing it? <a href="https://docs.github.com/en/packages/using-github-packages-with-your-projects-ecosystem/configuring-dotnet-cli-for-use-with-github-packages" rel="noopener nofollow ugc">Configuring dotnet CLI for use with GitHub Packages</a> suggest setting these but overlooks the fact that <code>nuget.config</code> is needed locally and in source control and we don’t want the tokens committed to source control. That means hacks of dynamically creating / altering <code>nuget.config</code> or creating a duplicate CI templated version that can be set dynamically in Actions with repo secrets.</p>

**Comment by jcansdale** [comment, marked as answer]
<p>Hi <a class="mention" href="/ghudik-cas">@ghudik-cas</a>,</p>
<blockquote>
<p>However I thought that wouldn’t be required within the context of GitHub actions in the same org?</p>
</blockquote>
<p>The <code>${{ secrets.GITHUB_TOKEN }}</code> token is scoped to the workflow repository. What you can do is create an org level secret containing a PAT with the <code>read:packages</code> scope. You can then use something like <code>${{ secrets.READ_PACKAGES_TOKEN }}</code> from your workflow.</p>
<blockquote>
<p>I’m assuming I need to modify NuGet.config so it has the credentials like</p>
</blockquote>
<p>Almost. If you’re using .NET Core 3.1 or above, you can create a <code>nuget.config</code> (all lowercase) file at the base of your repository like:</p>
<pre><code class="lang-xml">&lt;?xml version="1.0" encoding="utf-8"?&gt;
&lt;configuration&gt;
    &lt;packageSources&gt;
        &lt;add key="github" value="https://nuget.pkg.github.com/OWNER/index.json" /&gt;
    &lt;/packageSources&gt;
    &lt;packageSourceCredentials&gt;
        &lt;github&gt;
            &lt;add key="Username" value="token" /&gt;
            &lt;add key="ClearTextPassword" value="%READ_PACKAGES_TOKEN%" /&gt;
        &lt;/github&gt;
    &lt;/packageSourceCredentials&gt;
&lt;/configuration&gt;
</code></pre>
<p>You can then set the <code>READ_PACKAGES_TOKEN</code> env var on your local machine and when restoring packages using your GitHub Actions workflow.</p>
<p>You could also create a <code>Directory.Build.targets</code> file to check that <code>READ_PACKAGES_TOKEN</code> actually exists.</p>
<p>I hope that helps!</p>

**Reply by igiona** [reply]
@jcansdale thanks for the suggestions.
After hours of digging into the poor documentation, at least here I find out that GITHUB_TOKEN (advertised in the official documentation as the way to go) does NOT work outside of the workflow repo.

Your suggestion of using a READ token works for the read... but for it prevents the ability to push, even if the api-key is set (which I guess is what you meant below with "just suppress the warning")

What would be the suggest approach, for the very strange usecase (ironic) of downloading AND uploading NuGet packages from/to the GitHub registry ? A read/write token (classic) set from a CI-service-user?

Thank you 

OT: For the readers that land here with issues with the GitHub registry, you might be missing
```
permissions:
  packages: write
```
in your workflow, without this permission GITHUB_TOKEN  will not be able to push.

**Comment by ghost** [comment]
<p>Thanks <a class="mention" href="/jcansdale">@jcansdale</a>. That helps. Why does it need to be an org level secret though? I thought that was only if you needed to not duplicate secrets between repos (just being used in one repo here). Also org level secrets require Admin I believe which most devs won’t have so it’s a bit more hassle to create having to route through an Admin.</p>
<p>As I mentioned I had also tried replacing <code>${{ secrets.GITHUB_TOKEN }}</code> with a token that has proper package access but the result was the same. Perhaps there was something else at play or the fact that it wasn’t an org level secret.</p>
<p>The restore is currently working (after populating nuget.config creds) but the push is failing; at the moment I’m using one token with read/write packages (and repo) for both the auth and restore as well as the push but the push result is <code>warn : *** does not have the correct permissions to execute 'CreatePackageVersion'</code>. It also has SSO enabled.</p>
<p>For the time being I’ve removed <code>actions/setup-dotnet</code> and my workflow looks like this:</p>
<pre><code class="lang-yaml">name: 'Build, Pack, Push'

on:
  workflow_dispatch:
  push:
    branches:
    - develop
  pull_request:
    branches:
    - develop

env:
  DOTNET_CLI_TELEMETRY_OPTOUT: true

jobs:
  build-push:
    runs-on: ubuntu-latest

    steps:
      -
        name: Prep
        run: |
          echo "::set-env name=BUILD_VER::2.3.$GITHUB_RUN_NUMBER"
          echo "::set-env name=CONFIG::Release"
          echo "::set-env name=NUGET_URL::https://nuget.pkg.github.com/${{ github.repository_owner }}/index.json"
          echo "::set-env name=SLN::MyApp.sln"

      -
        name: Checkout
        uses: actions/checkout@v2

      -
        name: Adjust NuGet Config Credentials
        run: |
          rm -f ./nuget.config
          mv nuget.ci.config nuget.config
          echo $(sed -e "s/\${USER}/${{ secrets.PACKAGE_USER }}/" -e "s@\${PAT}@${{ secrets.PACKAGE_TOKEN }}@" nuget.config) &gt;nuget.config

      -
        name: Build
        run: |
          dotnet build \
            ${{ env.SLN }} \
            /p:VersionPrefix=${{ env.BUILD_VER }} \
            --configuration ${{ env.CONFIG }}

      # Sadly we can't use a glob pattern with NuGet pack like we could in Azure DevOps
      -
        name: NuGet Pack
        run: |
          PROJECTS=**/*.csproj

          shopt -s nullglob
          shopt -s globstar

          for p in $PROJECTS
          do
            echo "Packing $p"
            dotnet pack $p \
              --configuration ${{ env.CONFIG }} \
              --no-build \
              /p:Version=${{ env.BUILD_VER }}
          done

      -
        name: NuGet Push
        run: |
          dotnet nuget list source
          dotnet nuget push **/*.nupkg \
            --source ${{ env.NUGET_URL }} \
            --api-key ${{ secrets.PACKAGE_TOKEN }} \
            --skip-duplicate
</code></pre>
<p>In any event the GH documentation around this area seems very lacking IMO but maybe I’m missing something.</p>

**Comment by jcansdale** [comment]
<blockquote>
<p>Why does it need to be an org level secret though? I thought that was only if you needed to not duplicate secrets between repos (just being used in one repo here).</p>
</blockquote>
<p>You’re right, it doesn’t need to be an org level secret. I find doing it this way convenient for the <code>read:packages</code> PAT. This way workflows have easy access to all of the orgs packages.</p>
<blockquote>
<p>I’m using one token with read/write packages (and repo) for both the auth and restore as well as the push but the push result is  <code>warn : *** does not have the correct permissions to execute</code></p>
</blockquote>
<p>The access is a combination of the PAT scopes and the read/write access the PAT’s owner has to the associated repository. Could you check that the owner of the PAT has write access to the repository associated with the package?</p>
<blockquote>
<p>my workflow looks like this</p>
</blockquote>
<p>Here are a few minor suggestions.</p>
<pre><code class="lang-auto">env:
  DOTNET_CLI_TELEMETRY_OPTOUT: true
</code></pre>
<p>You can use the following to quieten down dotnet (I <em>think</em> this includes telemetry opt out).</p>
<pre><code class="lang-auto">env:
  DOTNET_NOLOGO: true
</code></pre>
<p>Unfortunately this is ignored. All it will do is suppress the warning, which is misleading.</p>
<pre><code class="lang-auto">--api-key ${{ secrets.PACKAGE_TOKEN }}
</code></pre>
<p>Could you let me know the name of the repository this is in so I can investigate further? You can send a message using this link:<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <a href="https://support.github.com/request" target="_blank" rel="noopener nofollow ugc">support.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="" height="">

<h3><a href="https://support.github.com/request" target="_blank" rel="noopener nofollow ugc">GitHub Support</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Thanks,<br>
Jamie.</p>

**Comment by ghost** [comment]
<p><a class="mention" href="/jcansdale">@jcansdale</a>:</p>
<ul>
<li>Thanks for the cleanup tips - implemented.</li>
<li>I sent you a private message though I figured out the problem immediately after (of course)</li>
<li>It turns out I forgot to add the Team to the repo that the service account user is in.</li>
</ul>
<p>Small forgotten thing that wasn’t obvious to me from original error messages.</p>
<p>Thanks for the help</p>

**Comment by gautelo** [comment]
I also found it a bit strange that it was hard to glean from the main documentation how this was supposed to be set up.

I also had the requirement that I wanted to both read and write to both nuget.org and GH Packages. The solution that worked for me was to create a classic PAT with read/write access to packages for the orgs public repos. In the pipeline, rather than using a NuGet.Config file, I reconfigured the sources like this:

```yaml
      - name: 'NuGet: Configure sources'
        run: |
          dotnet nuget remove source nuget.org
          dotnet nuget add source -n nuget.org -u NUGETUSER -p ${{ secrets.NUGET_TOKEN }}    --store-password-in-clear-text "https://api.nuget.org/v3/index.json"
          dotnet nuget add source -n github    -u GH_ORG    -p ${{ secrets.NUGET_GH_TOKEN }} --store-password-in-clear-text "https://nuget.pkg.github.com/GH_ORG/index.json"
```

This way the `dotnet restore` works with retrieving packages from both feeds, and I can push normally using something like this:

```yaml
      - name: 'Nuget: Push to GH Packages'
        run: dotnet nuget push ./output/nuget/*.nupkg --source "github" --api-key ${{ secrets.NUGET_GH_TOKEN }}

      - name: 'Nuget: Push release to nuget.org'
        run: dotnet nuget push ./output/nuget/*.nupkg --source "nuget.org" --api-key ${{ secrets.NUGET_TOKEN }}
```

In my case I also have condition for running each of those, but that's besides the point of this post.

Hope this helps someone or that somebody from GitHub sees this and adds it to the docs.

**Reply by gautelo** [reply]
I've no idea why you have to pass the token both when defining the source and when using the source, just another quirk you'll need to know, I guess. :confused:
